File: upack/Command.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.IO;
using System.IO.Compression;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Reflection;
using System.Runtime.Serialization.Json;
using System.Text;
using System.Threading.Tasks;

namespace Inedo.ProGet.UPack
{
    public abstract class Command
    {
        [AttributeUsage(AttributeTargets.Property, AllowMultiple = false, Inherited = true)]
        public sealed class PositionalArgumentAttribute : Attribute
        {
            public int Index { get; }
            public bool Optional { get; set; } = false;

            public PositionalArgumentAttribute(int index)
            {
                this.Index = index;
            }
        }

        [AttributeUsage(AttributeTargets.Property, AllowMultiple = false, Inherited = true)]
        public sealed class ExtraArgumentAttribute : Attribute
        {
            public bool Optional { get; set; } = true;
        }

        public abstract class Argument
        {
            protected readonly PropertyInfo p;

            internal Argument(PropertyInfo p)
            {
                this.p = p;
            }

            public abstract bool Optional { get; }
            public string DisplayName => p.GetCustomAttribute<DisplayNameAttribute>()?.DisplayName ?? p.Name;
            public string Description => p.GetCustomAttribute<DescriptionAttribute>()?.Description ?? string.Empty;
            public object DefaultValue => p.GetCustomAttribute<DefaultValueAttribute>()?.Value;

            public abstract string GetUsage();

            public virtual string GetHelp()
            {
                return $"{this.DisplayName} - {this.Description}";
            }

            public bool TrySetValue(Command cmd, string value)
            {
                if (p.PropertyType == typeof(bool))
                {
                    if (string.IsNullOrEmpty(value))
                    {
                        p.SetValue(cmd, true);
                        return true;
                    }
                    bool result;
                    if (bool.TryParse(value, out result))
                    {
                        p.SetValue(cmd, result);
                        return true;
                    }
                    Console.WriteLine($@"--{this.DisplayName} must be ""true"" or ""false"".");
                    return false;
                }

                if (p.PropertyType == typeof(string))
                {
                    p.SetValue(cmd, value);
                    return true;
                }

                if (p.PropertyType == typeof(NetworkCredential))
                {
                    if (string.IsNullOrWhiteSpace(value))
                    {
                        p.SetValue(cmd, null);
                        return true;
                    }

                    var parts = value.Split(new[] { ':' }, 2);
                    if (parts.Length != 2)
                    {
                        Console.WriteLine($@"--{this.DisplayName} must be in the format ""username:password"".");
                        return false;
                    }

                    p.SetValue(cmd, new NetworkCredential(parts[0], parts[1]));
                    return true;
                }

                throw new ArgumentException(p.PropertyType.FullName);
            }
        }

        public sealed class PositionalArgument : Argument
        {
            internal PositionalArgument(PropertyInfo p) : base(p)
            {
            }

            public int Index => p.GetCustomAttribute<PositionalArgumentAttribute>().Index;
            public override bool Optional => p.GetCustomAttribute<PositionalArgumentAttribute>().Optional;

            public override string GetUsage()
            {
                var s = $"«{this.DisplayName}»";

                if (this.Optional)
                {
                    s = $"[{s}]";
                }

                return s;
            }
        }

        public sealed class ExtraArgument : Argument
        {
            internal ExtraArgument(PropertyInfo p) : base(p)
            {
            }

            public override bool Optional => p.GetCustomAttribute<ExtraArgumentAttribute>().Optional;

            public override string GetUsage()
            {
                var s = $"--{this.DisplayName}=«{this.DisplayName}»";

                if (this.Optional)
                {
                    s = $"[{s}]";
                }

                if (p.PropertyType == typeof(bool) && this.DefaultValue == (object)false && this.Optional)
                {
                    s = $"[--{this.DisplayName}]";
                }

                return s;
            }
        }

        public string DisplayName => this.GetType().GetCustomAttribute<DisplayNameAttribute>()?.DisplayName ?? this.GetType().Name;
        public string Description => this.GetType().GetCustomAttribute<DescriptionAttribute>()?.Description ?? string.Empty;
        public IEnumerable<PositionalArgument> PositionalArguments => this.GetType().GetRuntimeProperties()
            .Where(p => p.GetCustomAttribute<PositionalArgumentAttribute>() != null)
            .Select(p => new PositionalArgument(p))
            .OrderBy(a => a.Index);

        public abstract Task<int> RunAsync();

        public IEnumerable<ExtraArgument> ExtraArguments => this.GetType().GetRuntimeProperties()
            .Where(p => p.GetCustomAttribute<ExtraArgumentAttribute>() != null)
            .Select(p => new ExtraArgument(p));

        public string GetUsage()
        {
            var s = new StringBuilder("upack ");

            s.Append(this.DisplayName);

            foreach (var arg in this.PositionalArguments)
            {
                s.Append(' ').Append(arg.GetUsage());
            }

            foreach (var arg in this.ExtraArguments)
            {
                s.Append(' ').Append(arg.GetUsage());
            }

            return s.ToString();
        }

        public string GetHelp()
        {
            var s = new StringBuilder("Usage: ");

            s.AppendLine(this.GetUsage()).AppendLine().AppendLine(this.Description);

            foreach (var arg in this.PositionalArguments)
            {
                s.AppendLine().Append(arg.GetHelp());
            }

            foreach (var arg in this.ExtraArguments)
            {
                s.AppendLine().Append(arg.GetHelp());
            }

            return s.ToString();
        }

        internal static async Task<PackageMetadata> ReadManifestAsync(Stream metadataStream)
        {
            var serializer = new DataContractJsonSerializer(typeof(PackageMetadata));
            return await Task.Run(() => (PackageMetadata)serializer.ReadObject(metadataStream));
        }

        internal static void PrintManifest(PackageMetadata info)
        {
            Console.WriteLine($"Package: {info.GroupAndName}");
            Console.WriteLine($"Version: {info.Version}");
        }

        internal static async Task UnpackZipAsync(string targetDirectory, bool overwrite, ZipArchive zipFile, bool perserveTimestamps)
        {
            Directory.CreateDirectory(targetDirectory);

            var entries = zipFile.Entries.Where(e => e.FullName.StartsWith("package/", StringComparison.OrdinalIgnoreCase));

            int files = 0;
            int directories = 0;

            foreach (var entry in entries)
            {
                var targetPath = Path.Combine(targetDirectory, entry.FullName.Substring("package/".Length).Replace('/', Path.DirectorySeparatorChar));

                if (entry.FullName.EndsWith("/"))
                {
                    Directory.CreateDirectory(targetPath);
                    directories++;
                }
                else
                {
                    Directory.CreateDirectory(Path.GetDirectoryName(targetPath));
                    using (var entryStream = entry.Open())
                    using (var targetStream = new FileStream(targetPath, overwrite ? FileMode.Create : FileMode.CreateNew, FileAccess.Write, FileShare.None))
                    {
                        await entryStream.CopyToAsync(targetStream);
                    }

                    // Assume files with timestamps set to 0 (DOS time) or close to 0 are not timestamped.
                    if (perserveTimestamps && entry.LastWriteTime.Year > 1980)
                    {
                        File.SetLastWriteTimeUtc(targetPath, entry.LastWriteTime.DateTime);
                    }

                    files++;
                }
            }

            Console.WriteLine($"Extracted {files} files and {directories} directories.");
        }

        internal static async Task CreateEntryFromFileAsync(ZipArchive zipFile, string fileName, string entryPath)
        {
            var entry = zipFile.CreateEntry(entryPath);

            using (var input = new FileStream(fileName, FileMode.Open, FileAccess.Read))
            using (var output = entry.Open())
            {
                await input.CopyToAsync(output);
            }
        }

        internal static async Task CreateEntryFromStreamAsync(ZipArchive zipFile, Stream file, string entryPath)
        {
            var entry = zipFile.CreateEntry(entryPath);
            
            using (var output = entry.Open())
            {
                file.Position = 0;
                await file.CopyToAsync(output);
            }
        }

        internal static async Task AddDirectoryAsync(ZipArchive zipFile, string sourceDirectory, string entryRootPath)
        {
            bool hasContent = false;

            foreach (var fileName in Directory.EnumerateFiles(sourceDirectory))
            {
                await CreateEntryFromFileAsync(zipFile, fileName, entryRootPath + Path.GetFileName(fileName));
                hasContent = true;
            }

            foreach (var directoryName in Directory.EnumerateDirectories(sourceDirectory))
            {
                hasContent = true;
                await AddDirectoryAsync(zipFile, directoryName, entryRootPath + Path.GetFileName(directoryName) + "/");
            }

            if (!hasContent)
                zipFile.CreateEntry(entryRootPath);
        }

        internal static async Task<string> GetVersionAsync(string source, string group, string name, string version, NetworkCredential credentials, bool prerelease)
        {
            if (!string.IsNullOrEmpty(version) && !string.Equals(version, "latest", StringComparison.OrdinalIgnoreCase) && !prerelease)
            {
                return version;
            }

            using (var client = CreateClient(credentials))
            using (var response = await client.GetAsync($"{source.TrimEnd('/')}/packages?group={Uri.EscapeDataString(group)}&name={Uri.EscapeDataString(name)}"))
            {
                response.EnsureSuccessStatusCode();

                var serializer = new DataContractJsonSerializer(typeof(RemotePackageMetadata));
                var metadata = (RemotePackageMetadata)serializer.ReadObject(await response.Content.ReadAsStreamAsync());
                var versions = metadata.Versions.Select(UniversalPackageVersion.Parse);

                if (!prerelease)
                {
                    versions = versions.Where(v => string.IsNullOrEmpty(v.Prerelease));
                }

                return versions.Max().ToString();
            }
        }

        internal static HttpClient CreateClient(NetworkCredential credentials)
        {
            return new HttpClient(new HttpClientHandler
            {
                UseDefaultCredentials = credentials == null,
                Credentials = credentials,
                PreAuthenticate = true,
            });
        }
    }
}

File: command.go
package main

import (
	"archive/zip"
	"encoding/json"
	"fmt"
	"io"
	"io/ioutil"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"strings"
)

type Command interface {
	Name() string
	Description() string

	Help() string
	Usage() string

	PositionalArguments() []PositionalArgument
	ExtraArguments() []ExtraArgument

	Run() int
}

type PositionalArgument struct {
	Index       int
	Name        string
	Optional    bool
	Description string
	TrySetValue func(Command, *string) bool
}

type ExtraArgument struct {
	Name        string
	Required    bool
	Description string
	Flag        bool
	TrySetValue func(Command, *string) bool
}

func (a PositionalArgument) Help() string {
	return a.Name + " - " + a.Description
}

func (a ExtraArgument) Help() string {
	return a.Name + " - " + a.Description
}

func trySetBoolValue(name string, f func(Command) *bool) func(Command, *string) bool {
	return func(cmd Command, value *string) bool {
		if value == nil || *value == "" {
			*f(cmd) = true
			return true
		}

		if strings.EqualFold(*value, "true") {
			*f(cmd) = true
			return true
		}

		if strings.EqualFold(*value, "false") {
			*f(cmd) = false
			return true
		}
		fmt.Println("--"+name, "must be \"true\" or \"false\".")
		return false
	}
}

func trySetStringValue(name string, f func(Command) *string) func(Command, *string) bool {
	return func(cmd Command, value *string) bool {
		if value == nil {
			return false
		}

		*f(cmd) = *value
		return true
	}
}

func trySetBasicAuthValue(name string, f func(Command) **[2]string) func(Command, *string) bool {
	return func(cmd Command, value *string) bool {
		if value == nil || *value == "" {
			*f(cmd) = nil
			return true
		}

		parts := strings.SplitN(*value, ":", 2)
		if len(parts) != 2 {
			fmt.Println("--"+name, "must be in the format \"username:password\".")
			return false
		}

		*f(cmd) = &[2]string{parts[0], parts[1]}
		return true
	}
}

func (a PositionalArgument) Usage() string {
	s := "«" + a.Name + "»"

	if a.Optional {
		s = "[" + s + "]"
	}

	return s
}

func (a ExtraArgument) Usage() string {
	if !a.Required && a.Flag {
		return "[--" + a.Name + "]"
	}

	s := "--" + a.Name + "=«" + a.Name + "»"

	if !a.Required {
		s = "[" + s + "]"
	}

	return s
}

func defaultCommandUsage(cmd Command) string {
	s := []byte("upack ")

	s = append(s, cmd.Name()...)

	for _, arg := range cmd.PositionalArguments() {
		s = append(s, ' ')
		s = append(s, arg.Usage()...)
	}

	for _, arg := range cmd.ExtraArguments() {
		s = append(s, ' ')
		s = append(s, arg.Usage()...)
	}

	return string(s)
}

func defaultCommandHelp(cmd Command) string {
	s := []byte("Usage: ")

	s = append(s, cmd.Usage()...)
	s = append(s, '\n', '\n')
	s = append(s, cmd.Description()...)
	s = append(s, '\n')

	for _, arg := range cmd.PositionalArguments() {
		s = append(s, '\n')
		s = append(s, arg.Help()...)
	}

	for _, arg := range cmd.ExtraArguments() {
		s = append(s, '\n')
		s = append(s, arg.Help()...)
	}

	return string(s)
}

func ReadManifest(r io.Reader) (*PackageMetadata, error) {
	var meta PackageMetadata
	err := json.NewDecoder(r).Decode(&meta)
	if err != nil {
		return nil, err
	}
	return &meta, nil
}

func PrintManifest(info *PackageMetadata) {
	fmt.Println("Package:", info.groupAndName())
	fmt.Println("Version:", info.Version)
}

func UnpackZip(targetDirectory string, overwrite bool, zipFile *zip.Reader) error {
	err := os.MkdirAll(targetDirectory, 0777)
	if err != nil {
		return err
	}

	var files int
	var directories int

	for _, entry := range zipFile.File {
		if !strings.HasPrefix(strings.ToLower(entry.Name), "package/") {
			continue
		}

		targetPath := filepath.Join(targetDirectory, entry.Name[len("package/"):])

		if entry.Mode().IsDir() {
			err = os.MkdirAll(targetPath, entry.Mode())
			if err != nil {
				return err
			}
			directories++
		} else {
			err = os.MkdirAll(filepath.Dir(targetPath), entry.Mode())
			if err != nil {
				return err
			}
			err = saveEntryToFile(entry, targetPath, overwrite)
			if err != nil {
				return err
			}

			files++
		}
	}

	fmt.Println("Extracted", files, "files and", directories, "directories.")
	return nil
}

func saveEntryToFile(entry *zip.File, targetPath string, overwrite bool) (err error) {
	r, err := entry.Open()
	if err != nil {
		return
	}
	defer func() {
		if e := r.Close(); err == nil {
			err = e
		}
	}()

	flags := os.O_WRONLY | os.O_TRUNC | os.O_CREATE
	if !overwrite {
		flags |= os.O_EXCL
	}

	f, err := os.OpenFile(targetPath, flags, entry.Mode())
	if err != nil {
		return
	}
	defer func() {
		if e := f.Close(); err == nil {
			err = e
		}
	}()

	_, err = io.Copy(f, r)
	return
}

func CreateEntryFromFile(zipFile *zip.Writer, fileName, entryPath string) (err error) {
	f, err := os.Open(fileName)
	if err != nil {
		return
	}
	defer func() {
		if e := f.Close(); err == nil {
			err = e
		}
	}()

	fi, err := f.Stat()
	if err != nil {
		return
	}

	h, err := zip.FileInfoHeader(fi)
	if err != nil {
		return
	}

	h.Name = entryPath
	h.Method = zip.Deflate

	w, err := zipFile.CreateHeader(h)
	if err != nil {
		return
	}

	_, err = io.Copy(w, f)
	return
}

func CreateEntryFromStream(zipFile *zip.Writer, file io.Reader, entryPath string) (err error) {
	w, err := zipFile.Create(entryPath)
	if err != nil {
		return
	}

	_, err = io.Copy(w, file)
	return
}

func AddDirectory(zipFile *zip.Writer, sourceDirectory, entryRootPath string) (err error) {
	fi, err := os.Stat(sourceDirectory)
	if err != nil {
		return
	}

	h, err := zip.FileInfoHeader(fi)
	if err != nil {
		return
	}

	h.Name = entryRootPath

	_, err = zipFile.CreateHeader(h)
	if err != nil {
		return
	}

	infos, err := ioutil.ReadDir(sourceDirectory)
	if err != nil {
		return
	}
	for _, fi := range infos {
		if fi.IsDir() {
			err = AddDirectory(zipFile, filepath.Join(sourceDirectory, fi.Name()), entryRootPath+fi.Name()+"/")
		} else {
			err = CreateEntryFromFile(zipFile, filepath.Join(sourceDirectory, fi.Name()), entryRootPath+fi.Name())
		}

		if err != nil {
			return
		}
	}

	return
}

func GetVersion(source, group, name, version string, credentials *[2]string, prerelease bool) (string, error) {
	if version != "" && !strings.EqualFold(version, "latest") && !prerelease {
		return version, nil
	}

	req, err := http.NewRequest("GET", strings.TrimRight(source, "/")+"/packages?"+(url.Values{"group": {group}, "name": {name}}).Encode(), nil)
	if err != nil {
		return "", err
	}

	if credentials != nil {
		req.SetBasicAuth(credentials[0], credentials[1])
	}

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		return "", fmt.Errorf("ProGet returned HTTP error: %s", resp.Status)
	}

	var data RemotePackageMetadata
	err = json.NewDecoder(resp.Body).Decode(&data)
	if err != nil {
		return "", err
	}

	var latestVersion *UniversalPackageVersion
	for _, v := range data.Versions {
		version, err := ParseUniversalPackageVersion(v)
		if err != nil {
			return "", err
		}
		if !prerelease && version.Prerelease != "" {
			continue
		}
		if latestVersion == nil || latestVersion.Compare(version) < 0 {
			latestVersion = version
		}
	}
	return latestVersion.String(), nil
}
